core/src/main/java/org/apache/calcite/util/Util.java

```diff
@@ -544,6 +544,21 @@ public static void println(
     pw.println();
   }
 
+  /**
+   * Formats a double value to a String ensuring that the output
+   * is in scientific notation if the value is not "special".
+   * (Special values include infinities and NaN.)
+   */
+  public static String toScientificNotation(Double d) {
```

Comment:
Should this method guarantee that there is a unique string for every possible `double` value? (Except for the many representations of `NaN` etc.) And that it has an inverse method that will convert the string to the same `double` value.

I believe that we are relying on this bijection, because there is no direct way to specify a binary floating point value in SQL.

Reply:
Unfortunately there is no bijection. This conversion is lossy in both directions, from String to Double and from Double to String. This method is not 100% adequate for generating SQL, but it is adequate for e.g., displaying plans. 

However, the lossy conversion problem already exists in Calcite and is not specific to DOUBLE values: see https://issues.apache.org/jira/browse/CALCITE-5987. I have discussed in a message on the dev list about how this separate issue can be tackled.

Reply:
I realize that my previous message is wrong, indeed, there is no way to specify a binary floating point value in other SQL dialects. I don't think we can solve this problem, but we didn't introduce this problem in this PR. The problem exists already if one allows simplifications that can produce FP results.